doc: Make PR description a contributor expectation (#75115)
zephyrproject-rtos/zephyr: doc/contribute/contributor_expectations.rst

- Each commit in the PR must provide a commit message following the
  :ref:`commit-guidelines`.

> - The PR must include a description that summarizes the changes in the PR

Reviewer: nice,  but do you think people will read this? Still good to get in, but IMO we need to enforce this in CI, fail PRs without any description.

Reviewer 2: Don't we do it already?

Reviewer 2: Nevermind, PR description, not commit message -- yeah I agree ideally the CI would check this.

Reviewer 2: https://github.com/zephyrproject-rtos/zephyr/pull/75788

>   by describing what the changes are and why they are being made.

Reviewer 2: I would swap this, the important part is the "why", the "what" you can tell from the code itself (hopefully). How about "by describing why the change is needed and what the change is"?

Author: Personally disagree, that wording sounds awkward to my ears. Also, I think the 'what' is actually the main motivation for this PR, because it is annoying to see a PR with no description and try to figure out from the code what it's doing. The 'why' is only important to help form a judgement about the PR, but the first step is understanding the 'what'

Reviewer 2: Well the "what" should be already covered by the subject. Anyway, both are mentioned (and yeah it would sound awkward) so fine by me either way.

Reviewer 2: It depends on the change, ideally the code is perfectly understandable and does not need a verbal description and the oneliner tells everything you need to know to review the PR. Reality is different, obviously.

Reviewer 3: I disagree. We need more than a one-liner, which is also why we require a commit message _body_.

Reviewer 2: Sorry I wasn't trying to say that we need only the oneliner. What I wanted to say is that the oneliner should already describe what the change is, and the body should expand on that adding the motivation for doing the change.

Author: Maybe for a small PR like less than 30 lines you can tell from the code, but I am not uncommonly seeing hundreds/thousands line change PRs with no description. 

When writers/presenters want to convey some information, especially in a technical forum, they put an introduction and conclusion. Why? If the body already contains all the information what is the point of the introduction and conclusion. Well, it's because human brains are not computers (both the writers and the readers), and without some more explicit framing of the content, even if said content is in written form, there may be some misunderstanding or things missed about what is being conveyed. In school, I always heard something like "tell them what you are going to tell them, tell them, then tell them what you just told them" from professors speaking of technical communication. I think a similar sentiment should and in fact DOES apply to an open source software PR to a large extent, because it's also effectively about conveying information to an audience.

Reviewer 2: Let me step back for a sec:

I am NOT (I repeat: NOT) saying that we should allow PR with empty description. Our compliance check requires a description, if it wasn't requiring a description, I would change it to require a description, if we did not have compliance requiring a description, I would ask contributors to add a description. I love descriptions, hopefully that leaves no ambiguity.

That said, if you feel like you NEED the description to understand the change, maybe the change itself is not good enough. Maybe it should be refactored, maybe it should be split, maybe it's absolutely perfect and it's just intrinsically complex and it needs to be explained.

You may see some very understandable code and not even check the commit message and get it in. Or you may see a very badly written change with a perfectly written description, you would block it asking to refactor the code.

Reviewer 2: > I am not uncommonly seeing hundreds/thousands line change PRs with no description

Again the compliance check requires a description, you see this in the project, the compliance check is broken and we need to fix it. So if that's the case, point to an example please.

Reviewer 2: Ah, @henrikbrixandersen kindly pointed out to me that this is referring to the PR description and somehow I was thinking about the commit message. So yeah scrape half of what I said -- yes ideally the CI would flag empty PR messages.

> - The PR must include a description that summarizes the changes in the PR
>   by describing what the changes are and why they are being made.

Reviewer 4: there is a repetition in the sentence that I think is not ideal (...PR must include a description of the changes that describes the changes ...)

```suggestion
- The PR's description must include a summary of the changes and their rationale.
```

Author: updated
